### ISSUE TITLE
cannot use NVDA internal dictionary with IBMTTS

### ISSUE THREAD

**Issue by PlatinumTsuki** [ultrasound1372/NVDA-IBMTTS-Driver#11]
in the old version of eloquence ... you could switch between using NVDA's internal dictionary or not, in IBMTTS this feature has disappeared.

**Comment by Mohamed00** [comment]
Internal dictionary? I've only ever seen this option once in a French 
version. I don't think it ever made it to the most common builds.

**Comment by PlatinumTsuki** [comment]
indeed, I too discovered it with a French version of the extension, when the function is active ... it clearly improves the pronunciation of the French language.  and so ... you're telling me it's impossible to include it in IBMTTS?

**Comment by Mohamed00** [comment]
It might be a possibility, however it would be nice to possibly find the 
author and ask them for permition to integrate their code first. Most of 
the word replacements in the internal dictionary seem to be more 
appropriate for Eloquence's built-in main and root dictionaries, though.

**Comment by PlatinumTsuki** [comment]
it would be really good if you could implement it in IBMTTS, being a french speaking user of the extension ... it will make my life much easier, in any case if I can do something to help ... ask me  and I will do it with great pleasure.

**Comment by ultrasound1372** [comment]
I need you to explain to me what you mean. What are you calling the internal dictionary? NVDA's builtin.dic? Eloquence root dicts? If eloquence root/main dicts, I think with the latest version which will be released fairly soon it will only load the dictionaries made for English if using the English settings, but I might be wrong. If I am wrong, I'll have to see how I may implement that, because the dictionary loading is present in the lower level driver which I don't think is aware which language NVDA wants. I have been told though that if you load it up on English mode and then switch to another language that ECI is smart enough to not apply the dictionaries to the wrong language, but I'm not sure how that is if you initialize in something other than English. I'll give that a look.

**Comment by Mohamed00** [comment]
This was a dictionary that was built into a French addon. There was a 
dedicated .py file with all the entries, and there was an option in the 
user interface to enable or disable it. It fixed several issues related 
to French, such as a few words that Eloquence would sometimes mistake 
for verbs, time pronunciation, and several words. However, when I looked 
at it, it seemed that many of the entries would be suited for a French 
root dictionary if support for non-English main and root dictionaries 
gets implemented.

**Comment by ultrasound1372** [comment]
It's supposed to be implemented, but currently only on initialization since that's two different levels of the synth driver. Given this, some tweaks some people want to do that you can't just do in eloquence dictionaries but requiring tags such as the code in YATES, I may implement a sort of py file embed exec feature to load a python file from the eci libraries directory, and do one of two things. Either it expects specially formatted dictionaries from that py file with a mapping of languages to dictionary contents to be resubed, or it calls a function in that file giving a string of text (after backquote suppression has been applied) and a language code of some sort, and then expects a string to come back out of that function and just treats that as the new text. Since we can't implement filters in ECI itself. If an exception were raised while calling this function however it would just leave the input string unchanged and proceed to do a few of it's own calculations like pause and time expressions, forcing head parameters, and then pass to synthesis. Might that be a reasonable solution to this? That would let us in one way or another incorperate this py file and some of the changes from YATES, and could be expanded in the future.
There are of course security implications for just blindly running code like this. Another approach may be some sort of dictionary file that is able to store dictionaries for multiple languages, maybe in JSON format? That way there's no execution, just pure data, except of course the regular expressions we'd be compiling. This would just kind of expect the data to be regexes.

**Comment by PlatinumTsuki** [comment]
I'm really sorry: I can't follow the conversation properly because it's all too complicated for me,
nevertheless ... with a quick inspection on my part of the old version of eloquence, I can easily confirm Mohamed00's statements regarding the .py file of the internal dictionary,
here is a direct link to the French version of eloquence which contains this famous dictionary ... I leave it to you to inspect it deeply to understand what it is all about:
https://www.nvda-fr.org/voix/EloquenceAutoLanguageSwitching.nvda-addon
I hope it can help you

**Comment by PlatinumTsuki** [comment]
excuse me for the inconvenience ... but I would like to know if there is any news concerning my request?

**Comment by Mohamed00** [comment]
Coming back to this. @ultrasound1372, since we now have full support for multilingual root and main dictionaries, perhaps it may be appropriate to revisit this? Perhaps we could merge the basic expressions into french_fixes, and the words into fraroot.dic? Though actually, thinking about it again, it's probably not a good idea to merge all this code since we don't fully understand what it does.

**Comment by ultrasound1372** [comment]
It's quite an undertaking to merge such a thing, and I don't know any French. It would have to be someone else's job to look at ECI's French phoneme table and merge in words that could go into root and main, and probably submit that to one of the community dictionaries. We just have to hope that most of the fixes will be put into one of those so the number of actual expressions we have to deal with will be sharply decreased, and then we can sort out ones that can be used in NVDA dictionaries.
